### PR TITLE
NDRS-161 - respect block limits

### PR DESCRIPTION
**Pull request by dwerner**
This starts the `BlockProposer` checking the `DeployConfig` values for proposed blocks.

Ref: 

https://casperlabs.atlassian.net/browse/NDRS-161
https://casperlabs.atlassian.net/browse/NDRS-295

### CODE REVIEW

File: node/src/components/block_proposer.rs
Comment (dwerner) [inline node/src/components/block_proposer.rs:400]: @goral09 and @afck can I please get a sanity check here? How would we check/enforce the size of the proposed block?

Comment (afck) [reply node/src/components/block_proposer.rs:400]: I agree that this requires serializing the block and all its deploys, and summing up the sizes. Obviously we don't have the serialized block here yet, so it's only about the length of the deploy hash list plus the sizes of the deploys (plus maybe some constant for the rest of the block?). We don't have to serialize the former since all deploy hashes have the same size. How to do the latter I don't know; maybe this should happen in another component (e.g. store each deploy's size when we receive it, before deserializing it)?

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:363]: Or:
```rust
if !deploy_config.max_payment_cost.is_zero()
```
That avoids allocating the zero bytes for the comparison.

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:368]: I have to admit this is the first time I see `max_payment_cost` — but isn't this about how much it costs the system to _execute_ the payment, and not about the maximum amount allowed to transfer in a single transaction?

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:417]: Isn't the gas price the number of _tokens per unit of gas_?
What we need to add up here is the "willingness to pay", i.e. the product of the gas price and the deploy's gas limit. (Maybe deploys don't have that yet?)

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:419]: This might just be a particularly big deploy — the next one might fit into the block again.
We'll need to sort the deploys:
* Highest gas _price_ comes first.
* For equal gas price, highest _gas limit_ comes first.

I believe we also want to allow transfers in the general section of the block: the one for the Wasm deploys.
So: The transfer part is reserved for transfers only, but transfers that outbid all remaining Wasm deploys should still get into the block, even outside of the transfer part.

(@AlexanderLimonov: Please correct me if I'm wrong!)

File: node/src/components/block_proposer.rs
Comment (goral09) [reply node/src/components/block_proposer.rs:368]: Yes, I think it is the latter. I don't know what it's part of the deploy config though. It should be in the contract runtime. @mpapierski @henrytill @EdHastingsCasperLabs can one of you confirm/deny it?

Comment (mpapierski) [reply node/src/components/block_proposer.rs:368]: @goral09 We don't currently support variable max payment costs on contract runtime. For now we have a hard coded `MAX_PAYMENT` value which we use https://github.com/CasperLabs/casperlabs-node/blob/41463933e2794e32c84581b6abc29e212fc30510/execution_engine%2Fsrc%2Fcore%2Fengine_state%2Fmod.rs#L98. On contract runtime it means the gas limit of a payment code, and also the minimum balance user is required to posses

File: node/src/components/block_proposer.rs
Comment (AlexanderLimonov) [reply node/src/components/block_proposer.rs:417]: Yes, we should both introduce gas reservations/limits (I prefer "reservation" because CEP#22 makes you pay for everything, not just what you use, refunds aside) if they're not there and check the sum of these against the block gas limit.

File: node/src/components/block_proposer.rs
Comment (AlexanderLimonov) [reply node/src/components/block_proposer.rs:419]: Yes, that's the correct ordering, the idea is to maximize revenue and to enable a coherent mapping from these deploy orderings to a "price," which, with reasonably small deploys dominating, would just be the marginal willingness to pay in the last block.

It seems I may not have written the provision for allowing overflowing WASMless transfers into the general section, but I believe that should be done (letting these transfers compete on price and volume as usual with regular deploys).

I think the final version of the ordering is this - lump *all* deploys together and sort (first by price, then by reservation), then go down the list and include each in their respective sections until WASMless transfers overflow, proceeding as normal until the regular deploy gas limit is reached.

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs:400]: From our sidebar on this, the end result is that we can calculate it from the ExecutableDeployItem and hashes.

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs:368]: Thanks, I've removed this check.

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs:417]: Per jira comment, we'll be addressing this after CEP22 merges.

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs:419]: Per jira comment, we'll be addressing this after CEP22 merges.

File: client/lib/executable_deploy_item_ext.rs
Comment (afck) [inline client/lib/executable_deploy_item_ext.rs:32]: Did you just flip those two (`_by_hash` and `_by_name`)?

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:394]: (Not necessarily in this PR, but:)
I thought we don't actually want a `max_transfers`, only a `max_deploys`. In other words: We want to reserve space for transfers, and use the remaining space for both transfers and deploys. (@AlexanderLimonov?)

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:401]: This should be a disjunction:
```rust
            if block_gas_running_total >= block_gas_limit
                || block_size_running_total >= max_block_size_bytes
                || (at_max_deploys && at_max_transfers)
```

Comment (afck) [reply node/src/components/block_proposer.rs:401]: But I wonder whether we should somehow limit the number of iterations where we don't add a deploy instead. Otherwise if we're one byte below the size limit, we will iterate through all the rest of `pending` and never add anything.

At least we could already `break` if:
```rust
block_gas_running_total + min_gas_per_deploy > block_gas_limit
|| block_size_running_total + min_size_per_deploy > max_block_size_bytes
```
(…or rather also with `checked_add`.)

File: node/src/components/block_proposer/event.rs
Comment (afck) [inline node/src/components/block_proposer/event.rs:33]: (It looks like this shouldn't be an enum at all: Both variants have the same fields. But it's a matter of taste.)

File: node/src/types/deploy.rs
Comment (afck) [inline node/src/types/deploy.rs:480]:
```suggestion
    /// Returns the `DeployType`.
```

File: node/src/types/deploy.rs
Comment (afck) [inline node/src/types/deploy.rs:496]: You mean we are _not_ a transfer?

File: node/src/types/deploy.rs
Comment (afck) [inline node/src/types/deploy.rs:500]: These two `expects` are not something a user can make fail, by putting e.g. a non-number or nothing into the "amount" argument, are they?

File: client/lib/executable_deploy_item_ext.rs
Comment (dwerner) [reply client/lib/executable_deploy_item_ext.rs:32]: Yeah, the ordering was different from trait to implementation.

File: node/src/types/deploy.rs
Comment (dwerner) [reply node/src/types/deploy.rs:496]: Nice catch, I left the comment out of date, thanks.

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs]: For the time being, I think we can't exit early based on `block_gas_limit` because (and this reminds me to mention it to @mpapierski :) ) transfers currently have a payment amount of 0, until this constant has been defined. We would exit early before checking for more (in this case zero-cost) transfers.

File: node/src/components/block_proposer.rs
Comment (AlexanderLimonov) [reply node/src/components/block_proposer.rs:398]: Regarding your last sentence - yes. I'm a bit confused about the limits imposed here being in terms of #s rather than gas (though, for transfers, it doesn't make a practical difference, only a difference for readability).

File: node/src/types/deploy.rs
Comment (dwerner) [reply node/src/types/deploy.rs]: It's a good point - I've mapped these to a new `deploy::Error` variant, and bubble that up to the validator, log it on on `error!` and returning `Effects::new()` from the event handler if this were to happen.

File: node/src/components/block_proposer.rs
Comment (mpapierski) [reply node/src/components/block_proposer.rs]: @dwerner I'm working on adding gas cost to wasmless transfers as part of https://casperlabs.atlassian.net/browse/EE-1122. That cost will come from EE's protocol data rather than being a constant defined somewhere. Would it be an issue if you'd have to query EE's protocol data to get wasmless transfer price here?

Comment (dwerner) [reply node/src/components/block_proposer.rs]: @mpapierski what would be the lifetime of that value? I'm going to guess that it shouldn't need to be queried very frequently... but I could be wrong, of course.

File: node/src/components/block_proposer/event.rs
Comment (dwerner) [reply node/src/components/block_proposer/event.rs:33]: For now I'd like to leave as is, but if this ends up remaining the case for long, I'd be ok to refactor this as you suggest.

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs:440]: Sorry this is when Gas overflows it's underlying datatype.

Comment (afck) [reply node/src/components/block_proposer.rs:440]: Ah, right, sorry!

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs:403]: Sure, I've logged it here: https://casperlabs.atlassian.net/browse/NDRS-669

File: node/src/components/block_proposer.rs
Comment (mpapierski) [reply node/src/components/block_proposer.rs]: > 
> @mpapierski what would be the lifetime of that value? I'm going to guess that it shouldn't need to be queried very frequently... but I could be wrong, of course.

That's right. This value won't change for given protocol version

**Review by henrytill — APPROVED** [review]
LGTM.  You might wait for another approval before merging.

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:53]:
```suggestion
/// Experimentally, deploys are in the range of 270-280 bytes, we use this to determine if we are
```

File: node/src/components/block_proposer.rs
Comment (afck) [reply node/src/components/block_proposer.rs:398]: So I guess this code will be updated accordingly when [CEP 22](https://github.com/CasperLabs/ceps/pull/22) gets implemented?

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:421]: The `continue` criterion above could be added to these three, too.

File: node/src/components/block_proposer.rs
Comment (afck) [inline node/src/components/block_proposer.rs:441]: If I submit a deploy with payment amount `U512::MAX`, won't that always cause us to `break` here, even if the block is still empty?
Would it make sense to just `continue` instead?

**Review by afck — APPROVED** [review]

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs:441]: I think that makes more sense, rather than allowing the proposer's behaviour to be affected by sending a max value.

File: node/src/components/block_proposer.rs
Comment (dwerner) [reply node/src/components/block_proposer.rs:421]: Sorry I don't understand what you mean here.

**Review by afck — APPROVED** [review]

**Pull request merged**
